### PR TITLE
memory value fixed conflicts

### PR DESCRIPTION
**Pull request by hesirui**
This is the follow up pull request to https://github.com/ElektraInitiative/libelektra/pull/2699
there were conflicts with my local branch, which I could not resolve with rebasing or merging, thus I created a new pull request.

## Basics

Check relevant points but **please do not remove entries**.
Do not describe the purpose of this PR in the PR description but:

- [x] Short descriptions should be in the release notes (added as entry in
      `doc/news/_preparation_next_release.md` which contains `_(my name)_`)
      **Please always add something to the the release notes.**
- [x] Longer descriptions should be in documentation or in design decisions.
- [x] Describe details of how you changed the code in commit messages
      (first line should have `module: short statement` syntax)
- [x] References to issues, e.g. `close #X`, should be in the commit messages.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [x] I added unit tests
- [x] I ran all tests locally and everything went fine
- [x] affected documentation is fixed
- [x] I added code comments, logging, and assertions (see doc/CODING.md)
- [x] meta data is updated (e.g. README.md of plugins and doc/METADATA.md)

## Review

Remove the line below and add the "work in progress" label if you do
not want the PR to be reviewed:

@markus2330 please review my pull request

## Merging

Please add the "ready to merge" label when the build server and you say
that everything is ready to be merged.

### CODE REVIEW

**Comment by hesirui** [comment]
It seems odd, that I the following error occurs in the mac build only. 
../src/plugins/memoryvalue/memoryvalue.c:107:31: error: format specifies type 'unsigned long' but the argument has type 'kdb_unsigned_long_long_t' (aka 'unsigned long long') [-Werror,-Wformat]
        snprintf (buf, n + 1, "%lu", normalizedMemVal);

when I change to long long %llu, the BSD build breaks.

**Comment by sanssecours** [comment]
> It seems odd, that the following error occurs in the mac build only.
> ../src/plugins/memoryvalue/memoryvalue.c:107:31: error: format specifies type 'unsigned long' but the argument has type 'kdb_unsigned_long_long_t' (aka 'unsigned long long') [-Werror,-Wformat]
> snprintf (buf, n + 1, "%lu", normalizedMemVal);
> 
> when I change to long long %llu, the BSD build breaks.

You can fix this problem using the macro [`ELEKTRA_UNSIGNED_LONG_LONG_F`](https://github.com/ElektraInitiative/libelektra/blob/master/src/include/kdbtypes.h#L122-L132) from [`kdbtypes.h`](https://github.com/ElektraInitiative/libelektra/blob/master/src/include/kdbtypes.h).

**Comment by hesirui** [comment]
ready to merge

File: doc/METADATA.ini [hunk 1]
```diff
@@ -1211,6 +1211,13 @@ description= check if value is a valid rgbcolor in hexformat
 # Old stuff inherited from filesys plugin
 #
 
+[check/memoryvalue]
```
Comment (markus2330) [inline doc/METADATA.ini]: "memoryvalue" is not really a well known term. Ideally you would rename it to `check/unit` and also allow other units, and not only "B".

File: doc/METADATA.ini [hunk 2]
```diff
@@ -1211,6 +1211,13 @@ description= check if value is a valid rgbcolor in hexformat
...
+type= empty
+status = implemented
+usedby/plugin= memoryvalue
+description= check if value is a valid memoryvalue
```
Comment (markus2330) [inline doc/METADATA.ini]: If you do not know what a memoryvalue is, this description does not help. An example would be very helpful.

File: src/plugins/memoryvalue/memoryvalue.c [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+	while ((cur = ksNext (returned)) != NULL)
+	{
+		const Key * meta = keyGetMeta (cur, "check/memoryvalue");
+		if (meta)
+		{
+			kdb_unsigned_long_long_t format = is_valid_key (cur);
+
+			if (format == 0)
+			{
+				return ELEKTRA_PLUGIN_STATUS_ERROR;
```
Comment (markus2330) [inline src/plugins/memoryvalue/memoryvalue.c]: No error set here.

File: src/plugins/memoryvalue/memoryvalue.c [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+#include "memoryvalue.h"
+#include <kdberrors.h>
+#include <kdbhelper.h>
+#include <kdbtypes.h>
+#include <regex.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+
+static kdb_unsigned_long_long_t is_valid_key (Key * key)
```
Comment (markus2330) [inline src/plugins/memoryvalue/memoryvalue.c]: Coding convention: isValidKey.

File: doc/METADATA.ini (hunk 1 above)
Comment (kodebach) [reply doc/METADATA.ini]: In that case this PR should be merged with #2614

File: src/plugins/memoryvalue/memoryvalue.c [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+{
+
+	const Key * meta = keyGetMeta (key, "check/memoryvalue");
```
Comment (kodebach) [inline src/plugins/memoryvalue/memoryvalue.c]: You already check that `check/memoryvalue` exists before calling this function and you don't use its value, so no need to check again.

File: src/plugins/memoryvalue/testmod_memoryvalue.c [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+	KeySet * ks = ksNew (20, KS_END);
+	// KeySet * ksGet = ks;
+
+	ksAppendKey (ks, hexkey);
+
+	PLUGIN_OPEN ("memoryvalue");
+
+	succeed_if ((plugin->kdbSet (plugin, ks, parentKey) >= 1 || errorExpected), "kdbSet did not succeed");
+
+	succeed_if ((plugin->kdbGet (plugin, ks, parentKey) >= 1 || errorExpected), "kdbGet did not succeed");
```
Comment (kodebach) [inline src/plugins/memoryvalue/testmod_memoryvalue.c]: You should call `kdbGet` then check that the value was correctly transformed (or an error occurred as expected). After that you should call `kdbSet` and check that the value was correctly restored.

You could also add some tests where the value is changed in between `kdbGet` and `kdbSet` and check that `kdbSet` doesn't overwrite the change.

File: src/plugins/memoryvalue/testmod_memoryvalue.c [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+#include <kdbconfig.h>
+#include <kdbtypes.h>
+#include <stdlib.h>
+#include <string.h>
+#include <tests_plugin.h>
+
+// the boolean parameter is needed, because we need unsigned longs as expected values, signed longs are too short thus no way to check for
+// invalid values with -1
+static void test_memoryvalue (const char * memoryvaluestring, kdb_unsigned_long_long_t memoryvalueexpected, bool errorExpected)
+{
```
[the remark this answers is not in the document]
Comment (markus2330) [reply src/plugins/memoryvalue/testmod_memoryvalue.c]: You can avoid the duplication by:

1. having a macro instead of a function (then the line numbers will be correct)
2. add the text what was tested to the assertions (which is more or less done here?)

**Comment by markus2330** [comment]
Btw. I am not unhappy with the functionality, so better we get a fully working plugin with the current functionality than a half-working plugin where

1. arbitrary units
2. kibibits/bytes

are supported.

File: src/plugins/memoryvalue/memoryvalue.c (hunk 3 above)
Comment (hesirui) [reply src/plugins/memoryvalue/memoryvalue.c]: When using the macro ELEKTRA_SET_ERRORF I get an error during compilation. Any Idea hot to fix that?

error:
      expected expression
                                ELEKTRA_SET_ERRORF (171, parentKey, "The value i...

File: src/plugins/memoryvalue/memoryvalue.c [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+static int elektraMemoryvalueConvertToByteString (Key * key, kdb_unsigned_long_long_t formatFactor)
+{
+
+	const char * str = keyString (key);
+	keySetMeta (key, "origvalue", str);
+	char * ptr;
+	kdb_unsigned_long_long_t ret;
+	kdb_unsigned_long_long_t normalizedMemVal;
+
+	ret = strtoll (str, &ptr, 10);
```
[the remark this answers is not in the document]
Comment (hesirui) [reply src/plugins/memoryvalue/memoryvalue.c]: I also get an error here if I use the macro.
error: use of
      undeclared identifier 'format'
                        ELEKTRA_UNSIGNED_LONG_LONG_S format

File: src/plugins/memoryvalue/memoryvalue.c (hunk 3 above)
Comment (markus2330) [reply src/plugins/memoryvalue/memoryvalue.c]: @kodebach can you help here?

File: src/plugins/memoryvalue/memoryvalue.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (hesirui) [reply src/plugins/memoryvalue/memoryvalue.c]: I forgot the import

File: src/plugins/memoryvalue/memoryvalue.c (hunk 3 above)
Comment (hesirui) [reply src/plugins/memoryvalue/memoryvalue.c]: Now It works I didnt import it.

**Comment by hesirui** [comment]
i fixed the problems above

**Comment by hesirui** [comment]
Unfortunately the tests break on the build server with the free BSD setup, on my local machine  (Mac) all tests pass.

File: doc/METADATA.ini (hunk 1 above)
Comment (markus2330) [reply doc/METADATA.ini]: In #2614 the author stopped with CM.

So this PR could use check/unit.

File: src/plugins/memoryvalue/memoryvalue.c [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+	while ((cur = ksNext (returned)) != NULL)
+	{
+		const Key * meta = keyGetMeta (cur, "check/memoryvalue");
+		if (meta)
+		{
+			kdb_unsigned_long_long_t format = isValidKey (cur);
+
+			if (format == 0)
+			{
+				ELEKTRA_SET_ERRORF (171, parentKey, "%s is not formatted properly!", keyString (cur));
```
Comment (markus2330) [inline src/plugins/memoryvalue/memoryvalue.c]: The error message could be improved. Please tell how it should be formatted and which key is affected.

File: src/plugins/memoryvalue/memoryvalue.c [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+
+	const char * value = keyString (key);
+	// else we manipulate the original
+	char * tempval = elektraStrDup (value);
+
+	kdb_unsigned_long_long_t ret;
+
+	char * endPtr;
+	// convert to long, if valid key, pointer should point to spaces or memory suffix like MB, GB
+	ret = ELEKTRA_UNSIGNED_LONG_LONG_S (tempval, &endPtr, 10);
```
Comment (markus2330) [inline src/plugins/memoryvalue/memoryvalue.c]: Check if this was successful? (Any number found?) Please also add a testcase with simply "B".

File: src/plugins/memoryvalue/README.md [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+
+## Introduction
+
+Memory Value Plugin checks the correct format of memory specifications.
+
+E.g. setting the max RAM consumption of JVM, 30MB would be valid while 30MBMB would be not valid.
+
+## Usage
+
+Add the metakey `check/memoryvalue` and set a memory value e.g. 20MB to the key.
```
Comment (markus2330) [inline src/plugins/memoryvalue/README.md]: Please improve the docu and describe how to use it (with examples, see other plugins).

File: src/plugins/memoryvalue/memoryvalue.c [hunk 12]
```diff
@@ -0,0 +1,188 @@
+/**
+ * @file
+ *
+ * @brief memoryvalue plugin
+ *
+ * @copyright BSD License (see LICENSE.md or https://www.libelektra.org)
+ *
+ */
+
+#define _XOPEN_SOURCE
```
Comment (markus2330) [inline src/plugins/memoryvalue/memoryvalue.c]: Why is this needed? Please add a comment.

File: src/plugins/memoryvalue/memoryvalue.c [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+#include "memoryvalue.h"
+#include <kdberrors.h>
+#include <kdbhelper.h>
+#include <kdbtypes.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+static char * deblank (char * input)
```
Comment (markus2330) [inline src/plugins/memoryvalue/memoryvalue.c]: Why are you returning the pointer if you modify the content of input itself? Please add a short docu what this function is doing.

File: src/plugins/memoryvalue/memoryvalue.c [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+	{
+		return 1;
+	}
+	else
+	{
+		return 0;
+	}
+}
+
+// formatFactor is used to determine by which factor the value has to multiplied to normalize to bytes
```
Comment (markus2330) [inline src/plugins/memoryvalue/memoryvalue.c]: Use @param.

File: src/plugins/memoryvalue/memoryvalue.c [hunk 15]
```diff
@@ -12,63 +12,63 @@
...
+static kdb_unsigned_long_long_t isValidKey (Key * key)
+{
 
-	if (compile_failure)
-	{
-		return 0;
-	}
+	const char * value = keyString (key);
+	// else we manipulate the original
+	char * tempval = elektraStrDup (value);
```
Comment (kodebach) [inline src/plugins/memoryvalue/memoryvalue.c]: You forgot to `elektraFree(tempval)`. If you don't modify the value (i.e. `deblank` just skips all spaces and returns the new position in the string) you won't have to `elektraStrDup` at all.

File: src/plugins/memoryvalue/memoryvalue.c (hunk 10 above)
Comment (hesirui) [reply src/plugins/memoryvalue/memoryvalue.c]: The problem with ELEKTRA_UNSIGNED_LONG_LONG_S, as I have understood it is, that the behavior is not defined in case of an error.

File: doc/METADATA.ini (hunk 1 above)
Comment (hesirui) [reply doc/METADATA.ini]: should I rename it to, check/unit?

File: src/plugins/memoryvalue/memoryvalue.c (hunk 10 above)
Comment (kodebach) [reply src/plugins/memoryvalue/memoryvalue.c]: `ELEKTRA_UNSIGNED_LONG_LONG_S` is just `strtoull(3)` or `strtoul(3)`, IIRC. Their behaviour in case of error is quite well defined.

Comment (hesirui) [reply src/plugins/memoryvalue/memoryvalue.c]: I know, but I thought it is not if the string specified is not valid.
http://www.cplusplus.com/reference/cstdlib/strtoull/

Comment (hesirui) [reply src/plugins/memoryvalue/memoryvalue.c]: Or should I just expect the string to be valid but simply not contain a number?
Then I can verfiy the conversion with 0ULL

**Comment by hesirui** [comment]
I fixed the remarked problems and renamed the plugin to check/unit.

**Comment by markus2330** [comment]
Thank you! Please make sure that the build server accepts it.

**Comment by hesirui** [comment]
I really don't know why the tests always fail here with the stroull approach, with regex it worked flawlessly. Also on my machine I dont get any error.

**Comment by markus2330** [comment]
@kodebach can you help here please?

**Comment by kodebach** [comment]
[These error messages](https://cirrus-ci.com/task/4660939836620800?command=tests#L201) seem pretty explicit to me. You initialized an unsigned integer to 0 and tried to decrement it.

**Comment by hesirui** [comment]
finally the build server accepts it

**Comment by markus2330** [comment]
Good to see this success! :+1: 

@kodebach is everything from your review now done?

@hesirui: the main Readme could still be improved, in particular with shell recorder tests (to have examples how your plugin can be used with the `kdb` tool). Maybe @sanssecours can help you there? As starting point you can take a look at https://master.libelektra.org/tests/shell/shell_recorder/tutorial_wrapper

**Comment by markus2330** [comment]
Thank you very much!
